src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,503 @@
+            //      var x = list[i]   or
+            //
+            // If so, we'll use those as the iteration variables for the new foreach statement.
+            var bodyStatements = GetBodyStatements(forStatement);
+            var (typeNode, foreachIdentifier, declarationStatement) = tryDeconstructInitialDeclaration();
+
+            if (typeNode == null)
+            {
+                // user didn't provide an explicit type.  Check if the index-type of the collection
+                // is different from than .Current type of the enumerator.  If so, add an explicit
+                // type so that the foreach will coerce the types accordingly.
+                var indexerType = GetIndexerType(containingType, collectionType);
+                if (!Equals(indexerType, iterationType))
+                {
+                    typeNode = (TTypeNode)generator.TypeExpression(
+                        indexerType ?? semanticModel.Compilation.GetSpecialType(SpecialType.System_Object));
+                }
+            }
+
+            // If we couldn't find an appropriate existing variable to use as the foreach
+            // variable, then generate one automatically.
+            if (foreachIdentifier.RawKind == 0)
+            {
+                foreachIdentifier = semanticFacts.GenerateUniqueName(
+                    semanticModel, forStatement, containerOpt: null, baseName: "v", cancellationToken);
+                foreachIdentifier = foreachIdentifier.WithAdditionalAnnotations(RenameAnnotation.Create());
+            }
+
+            // Create the expression we'll use to replace all matches in the for-body.
+            var foreachIdentifierReference = foreachIdentifier.WithoutAnnotations(RenameAnnotation.Kind).WithoutTrivia();
+
+            // Walk the for statement, replacing any matches we find.
+            findAndReplaceMatches(forStatement);
+
+            // Finally, remove the declaration statement if we found one.  Move all its leading
+            // trivia to the next statement.
+            if (declarationStatement != null)
+            {
+                editor.RemoveNode(declarationStatement,
+                    SyntaxGenerator.DefaultRemoveOptions | SyntaxRemoveOptions.KeepLeadingTrivia);
+            }
+
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            editor.ReplaceNode(
+                forStatement,
+                (currentFor, _) => this.ConvertForNode(
+                    (TForStatementSyntax)currentFor, typeNode, foreachIdentifier,
+                    collectionExpression, iterationType, options));
+
+            return document.WithSyntaxRoot(editor.GetChangedRoot());
+
+            // local functions
+            (TTypeNode, SyntaxToken, TStatementSyntax) tryDeconstructInitialDeclaration()
+            {
+                if (bodyStatements.Count >= 1)
+                {
+                    var firstStatement = bodyStatements[0];
```

Comment:
lots of this syntax fact stuff seems can be avoid if it uses IOperation

Reply:
potentially, though i don't know if it buys much.

IOperation is great when i want deep semantic information (especially around conversions and all the stuff not explicitly stated in code).  Sometimes, i actually just want to very specific common patterns between the languages.  I also want to not have to deal with anything implicit.  SyntaxFacts is really good for when you are basically doing something syntactic, but you want to abstract over VB/C#.

Reply:
IOperation has one more good thing if you bind multiple times. it is actually more efficient even if you keep same semantic model.

Reply:
I thought semantic model cached bound nodes internally.  regardless, i love ioperation.  But this is one of those cases where it doesn't seem to substantively improve the code.  

Note: i tried ioperation here.  I did not like it.  I found it more unweildy for this feature.